### ISSUE TITLE
What are the plans for threading / concurrency ?

### ISSUE THREAD

**Issue by aliraza-noon** [carp-lang/Carp#1149]
Hello there, really interesting project I stumbled on it through reddit post wondering what do you have in mind for concurrency and parallelism.

**Comment by aliraza-noon** [comment]
@eriksvedang that is really rude I guess you could have bothered to answer this with few words

**Comment by eriksvedang** [comment]
I'm very sorry, your account was almost empty and your comment quite generic, so I just assumed you were a bot.

**Comment by eriksvedang** [comment]
@aliraza-noon There is not a good and well though-out plan for threading and concurrency, I'm afraid.

**Comment by aliraza-noon** [comment]
ah ok I see I use this account for work.

In my previous life I have been somewhat involved in clojure community, my 2 cents maybe you can look at core.async which originated from go

**Comment by eriksvedang** [comment]
@aliraza-noon I see, sorry again. Yeah, I'd be very happy with something like core.async / Go! There has been experiments made with co-routines in Carp (though I can't find it now). Perhaps we can build something on top of that. Most likely anything like this will start as a library and we can add it to the sdtlib later.

**Comment by hellerve** [comment]
Might the experiment have been [gt.carp](https://github.com/hellerve/gt.carp)? It implements a system for cooperative coroutines connected with channels. Sadly the system isn’t quite sound yet, because the borrow checker cannot figure out when the function ends (since the system uses assembly to context switch, even in the middle of a function, sometimes forever).

**Comment by hellerve** [comment]
A little point for the future: I’d suggest asking question in the chat rather than opening an issue, just because questions generate a bit of noise for the maintainers. This is not a big issue, but we use the issue tracker to track problems and feature requests for the language. I suppose that this issue could be read as a feature request, but we try to keep it actionable by writing more comprehensive RFCs (see these as examples: #1115, #1095, #884).

**Comment by TimDeve** [comment]
I have written a small demo of using pthreads: https://github.com/TimDeve/pthread

**Comment by mingwugmail** [comment]
Hope threading and concurrency will be added soon to the language, or stdlib first.

Nowadays, any serious app need processing power on multi-cores.

**Comment by TimDeve** [comment]
@eriksvedang Correct me if I'm wrong but the borrow checker can't really handle threads safely at the moment so there would need to be some language change before something like that can be added to stdlib.

**Comment by hellerve** [comment]
The thing I proposed informally before is allowing for a special call that cleans up all references in a function before that function ends (basically the after-function cleanup, but as an explicit call). This would allow at least gt.carp to work, because the scope of a thread is always a Carp function.

**Comment by eriksvedang** [comment]
> @eriksvedang Correct me if I'm wrong but the borrow checker can't really handle threads safely at the moment so there would need to be some language change before something like that can be added to stdlib.

Yeah, I think that is correct. Now – it might not need that much changes to work. Definitely worth discussing more!

**Comment by TimDeve** [comment]
@hellerve I don't think it help with the issue I have in https://github.com/TimDeve/pthread, if you look at the example you have to be careful that the lifetime of your lambdas is longer than it takes for all your threads to finish, if the example wasn't joining before the lambdas are cleaned-up by the borrow checker you would get and use-after-free. We would lambda consumer to be aware of the lifetime of whatever the lambda captured. 
Unsafe example:
```
(load "git@github.com:TimDeve/pthread@v0.1.1")

(use-all Result Array)

(defn make-threads []
  (let-do [a-str &@"A string"
           callback &(fn [] (println* a-str))]
    (repeat 1000 &(fn [] (unsafe-from-success (Thread.create callback))))))

(defn main []
  (let-do [threads (make-threads)]
    (ignore
      (Thread.join-all threads))))
```

**Comment by hellerve** [comment]
This example feels a little wrong to me generally: I feel like the thread shouldn’t have a reference to the lambda here but own it (would mean to require copying the `fn`, but that makes sense to me, since otherwise the threads would share that memory). This way you ensure that the function only gets deleted when the thread gets deleted.

**Comment by TimDeve** [comment]
Even if you make it take ownership, maybe putting your callback in struct that's returned with your thread handle, you'll still have problems with the ref it captures, like this:
```clojure
(deftype Holder [f (Fn [] ())])

(defn make-thread []
   (let [r &(String.concat &[@"bro" @"ken"])
         ff (fn [] (println* r))]
     (Holder ff)))

(defn main []
  (let [t (make-thread)]
    (~(Holder.f &t))))
```
Output on my machine:
```
H]D
```

There needs to be a lifetime attached to lambdas that capture refs.

**Comment by eriksvedang** [comment]
> There needs to be a lifetime attached to lambdas that capture refs.

I think they have a lifetime, right? It's not checked properly though. I think we have an issue for this... just never got around to fix it.

**Comment by TimDeve** [comment]
Is the lifetime linked to the references it captures?

**Comment by eriksvedang** [comment]
> Is the lifetime linked to the references it captures?

That's the idea, though there is something missing for a certain situation iirc.

**Comment by redbar0n** [comment]
Golang's way of dealing with concurrency seems to be the gold standard nowadays. Interestingly, Go doesn't presume the programmer will deal with threads directly (which seems a wise choice, imho):

> Go blocks run your processes on a thread pool that contains a number of threads equal to two plus the number of cores on your machine, which means your program doesn’t have to create a new thread for each process. This often results in better performance because you avoid the overhead associated with creating threads.

https://www.braveclojure.com/core-async/

The downside is that it puts some extra weight on the shoulders of the language runtime. Namely a scheduler.

The issue with the `go` keyword + channels in Go, is that they don't afford simple local reasoning, black box abstraction, automatic error propagation, and reliable resource cleanup (according to [this article](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/)). In short: they are a bit too fine grained not contained, and need to be manually maintained. They might work well in imperative C-inspired languages such as Golang, but maybe not so much for languages aspiring to be more declarative and functional in style, such as Carp. Even though Clojure has their own way of dealing with them in [core.async](https://www.braveclojure.com/core-async/), they rely on an (imho unelegant) imperative [close!](https://clojuredocs.org/clojure.core.async/close!) keyword, which doesn't solve the problem of manual management and reliable resource cleanup.

So, another idea could be to build in and enforce something like a [fork/join](https://en.wikipedia.org/wiki/Verilog#Fork/join) construct, similar to Golang's [WaitGroup](https://pkg.go.dev/sync#WaitGroup). But more like [nurseries](https://trio.readthedocs.io/en/stable/reference-core.html#nurseries-and-spawning) in python's Trio library for concurrency, since nurseries allegedly alleviate the aforementioned problems. Basically, nurseries are a contained structure that will block the calling process until all the expressions within (which all execute concurrently) return. They afford the programmer with simple local reasoning and sequential reading of the program (as opposed to callbacks that [fragments the logic](https://youtu.be/drmNlZVkUeE?t=1845) and control flow out into various handlers). Plus, it shouldn't matter that the original process is blocked (waiting for their result), since the runtime will utilize available cores for the forked processes anyway.

For reference, here is the [hacker news criticism](https://news.ycombinator.com/item?id=16921761) of the aforementioned article on nurseries in Trio.

Without having thought thoroughly about it and explored the ramifications, I imagine such a nursery aka. fork/join construct could work well with borrowing checking and enable lifetime management, which could solve some problems mentioned previously in this discussion:

> cooperative coroutines connected with channels ... Sadly the system isn’t quite sound yet, because the borrow checker cannot figure out when the function ends

>the borrow checker can't really handle threads safely at the moment

>you have to be careful that the lifetime of your lambdas is longer than it takes for all your threads to finish

Since the construct will have a demarcated end point where all the forked processes will be done (since the main process is waiting for that).

But nurseries, [as far as I know](https://www.reddit.com/r/programming/comments/8es8x3/notes_on_structured_concurrency_or_go_statement/hu510gc/), work by mutating external variables (the [trio docs](https://trio.readthedocs.io/en/stable/tutorial.html#an-echo-client) corroborate that impression). But in a functional style we'd like the forked processes to work as expressions, returning a value.

One issue with fork/join constructs is: [how to order the results of each forked process](https://en.wikipedia.org/wiki/Verilog#Fork/join) at the moment they have completed and are supposed to be joined? An idea that might work here, is to automatically order the result of the forked processes according to their appearance in the calling code (the sequence in which they were forked).

Interested to hear what you think about this overall approach.

**Comment by dumblob** [comment]
@redbar0n the link "fragments the logic" points to a video which is unavailable:

>Video unavailable
>This video is no longer available due to a copyright claim by Alexander Miller

**Comment by redbar0n** [comment]
> @redbar0n the link "fragments the logic" points to a video which is unavailable:

Thanks. Fixed the link now.

**Comment by eriksvedang** [comment]
@redbar0n Thanks a lot for sharing your thoughts and research, that is super helpful!

**Comment by redbar0n** [comment]
This might be relevant reading: [The Scoped Task trilemma](https://without.boats/blog/the-scoped-task-trilemma/) (Apr 8, 2023).

Teaser excerpt:

> You want to use the Rust type system to implement an API for concurrent programming. Conceptually, in this API you have a “parent task” and one or more “child tasks;” if there is more than one child task, they can run concurrently with one another. Any sound API can only provide at most two of the following three desirable properties:
>   * Concurrency: Child tasks proceed concurrently with the parent.
>   * Parallelizability: Child tasks can be made to proceed in parallel with the parent.
>   * Borrowing: Child tasks can borrow data from the parent without synchronization.
